### PR TITLE
perf(ext/ffi): JIT compiler for FFI

### LINKED ISSUE #14431: FFI performance improvements (JIT compiler)
- [x] Value (de)serialization is slow. It has to go through JS -> `serde_json::Value` -> Rust, a faster path would be to convert to/fro a `serde_v8::Value` instead. https://github.com/denoland/deno/pull/14663
- [x] Each call has to go through the (already very thin) op layer. A possible solution is to make v8 callbacks during registration. This will get rid of the overhead of OpState. https://github.com/denoland/deno/pull/14945
- [x] JIT compiler for FFI calls. https://github.com/denoland/deno/pull/15006

### PR DESCRIPTION
**Pull request by littledivy**
This commit adds a tinycc-based JIT compiler
to generate optimized code to call into native code.

The allows performance equivalent to native C code. Here is what's happening:

image

image

Prior art:
- LuaJIT https://luajit.org/ext_ffi.html
- Bun FFI

Closes #14431 
Depends on https://github.com/denoland/rusty_v8/pull/1018

Co-Authored-By: @DjDeveloperr

### CODE REVIEW

File: bench.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+const lib = Deno.dlopen("./target/release/libtest_ffi.dylib", {
```
Comment (aapoalas) [inline bench.js:1]: Presumably this bench file should be removed before merging.

File: ext/ffi/jit.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+  };
+}
+
+fn native_to_c(ty: &NativeType) -> &'static str {
+  match ty {
+    NativeType::Void => "void",
+    NativeType::U8 => "unsigned char",
+    NativeType::U16 => "unsigned short",
+    NativeType::U32 => "unsigned int",
+    NativeType::U64 | NativeType::USize => "unsigned long",
```
Comment (aapoalas) [inline ext/ffi/jit.rs:23]: I'm definitely not an expert on C and had to google but internet seems to suggest that `int` and `long int` are both 4 bytes, whereas `long long` is 8 bytes. I have no idea if that's correct but I'll take the risk of asking an idiotic question: Are these correct?

Also, should these perhaps allow for 32bit systems?

File: ext/ffi/jit.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+macro_rules! impl_arg_int32 {
+  ($ctx: ident, $ty: ty) => {
+    paste::paste! {
+      unsafe extern "C" fn [<deno_ffi_ $ty>] (
+        info: *const v8::FunctionCallbackInfo,
+        index: i32,
+      ) -> $ty {
+        let scope = &mut v8::CallbackScope::new(&*info);
+        let info = v8::FunctionCallbackArguments::from_function_callback_info(info);
+        info.get(index).int32_value(scope).unwrap() as $ty
```
Comment (aapoalas) [inline ext/ffi/jit.rs:43]: I wonder what it takes to crash this? The algorithm for `int32_value` and `uint32_value` did not have errors for it, I think, but it starts with `Number(value)` which I think does have an error case.

But, I don't think that's exactly an issue as I think the error case is a supremely odd edge case.

File: ext/ffi/jit.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  ctx: Context,
+  c: String,
+}
+
+impl Compiler {
+  pub unsafe fn new() -> Result<Self, ()> {
+    let mut ctx = Context::new()?;
+    ctx.set_options(cstr!("-nostdlib"));
+
+    ctx.add_symbol(cstr!("deno_ffi_u8"), deno_ffi_u8 as *const c_void);
```
Comment (aapoalas) [inline ext/ffi/jit.rs:128]: This seems to be unnecessary, I think? The `impl_arg_uint32!(ctx, u8)` does this right?

File: ext/ffi/jit.rs [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+      .ctx
+      .add_symbol(cstr!("func"), sym.ptr.0 as *const c_void);
+
+    // extern <return_type> func(
+    self.c += "\nextern ";
+    self.c += native_to_c(&sym.result_type);
+    self.c += " func(";
+    // <param_type> p0, <param_type> p1, ...);
+    for (i, ty) in sym.parameter_types.iter().enumerate() {
+      if i > 0 {
```
Comment (aapoalas) [inline ext/ffi/jit.rs:164]: Possibly map + join?

File: ext/ffi/lib.rs [hunk 6]
```diff
@@ -628,40 +631,52 @@ where
...
-        Err(err) => {
-          deno_core::_ops::throw_type_error(scope, err.to_string());
-        }
-      };
-    },
-  )
-  .data(v8::External::new(scope, sym as *mut _).into())
-  .build(scope)
-  .unwrap();
+  let ctx = unsafe { crate::jit::Compiler::new() }.unwrap();
```
Comment (aapoalas) [inline ext/ffi/lib.rs:636]: Do these need error handling?

File: ext/ffi/lib.rs [hunk 7]
```diff
@@ -628,40 +631,52 @@ where
...
+    std::mem::transmute::<
+      *mut c_void,
+      extern "C" fn(*const v8::FunctionCallbackInfo),
+    >(alloc.addr)
+  };
+  let alloc = Box::into_raw(alloc);
+  let func = v8::FunctionBuilder::<v8::Function>::new_raw(func)
+    .build(scope)
+    .unwrap();
+  // let func = v8::Function::builder(
```
Comment (aapoalas) [inline ext/ffi/lib.rs:649]: Commented out code.

File: ext/ffi/tcc.rs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+#[derive(Debug)]
+pub struct TCCState {
+  _unused: [u8; 0],
+}
+pub const TCC_OUTPUT_MEMORY: i32 = 1;
+
+extern "C" {
+  pub fn tcc_new() -> *mut TCCState;
+  pub fn tcc_delete(s: *mut TCCState);
+  pub fn tcc_set_options(s: *mut TCCState, str: *const ::std::os::raw::c_char);
```
Comment (aapoalas) [inline ext/ffi/tcc.rs:21]: Lots of repeating of `::std::os::raw::`

File: ext/ffi/tcc.rs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    s1: *mut TCCState,
+    ptr: *mut ::std::os::raw::c_void,
+  ) -> ::std::os::raw::c_int;
+  pub fn tcc_get_symbol(
+    s: *mut TCCState,
+    name: *const ::std::os::raw::c_char,
+  ) -> *mut ::std::os::raw::c_void;
+}
+
+static GUARD: AtomicBool = AtomicBool::new(true);
```
Comment (aapoalas) [inline ext/ffi/tcc.rs:45]: Should this be `false` originally? It's being set to `true` on drop.

File: ext/ffi/jit.rs (hunk 2 above)
Comment (aapoalas) [reply ext/ffi/jit.rs:23]: Rust's `c_long` definition claims that on Windows this'll be 4 bytes but on Linux (often) 8 bytes.

File: ext/ffi/jit.c [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+void deno_rv_u16(void* info, unsigned short v);
+void deno_rv_u32(void* info, unsigned int v);
+void deno_rv_u64(void* info, unsigned long v);
+void deno_rv_i8(void* info, char v);
+void deno_rv_i16(void* info, short v);
+void deno_rv_i32(void* info, int v);
+void deno_rv_i64(void* info, long v);
+void deno_rv_f32(void* info, float v);
+void deno_rv_f64(void* info, double v);
+void deno_rv_pointer(void* info, void* v);
```
Comment (aapoalas) [inline ext/ffi/jit.c:25]: I guess the pointer calls are unimplemented yet?

File: ext/ffi/jit.rs (hunk 4 above)
Comment (littledivy) [reply ext/ffi/jit.rs:128]: Yes, very much WIP.

File: ext/ffi/tcc.rs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct TCCState {
+  _unused: [u8; 0],
+}
+pub const TCC_OUTPUT_MEMORY: i32 = 1;
+
+extern "C" {
+  pub fn tcc_new() -> *mut TCCState;
```
Comment (ry) [inline ext/ffi/tcc.rs:19]: where is tcc_new coming defined?

**Pull request closed without merging**